api/write.go

```diff
@@ -179,9 +179,15 @@ x:
 		select {
 		case batch := <-w.writeCh:
 			err := w.service.HandleWrite(context.Background(), batch)
-			if err != nil && w.errCh != nil {
-				w.errCh <- err
-			}
+			go func(err error) {
+				if err != nil && w.errCh != nil {
+					if len(w.errCh) == 0 {
+						w.errCh <- err
+					} else {
+						log.Warn("Cannot write error to error channel, it is not read")
+					}
+				}
+			}(err)
```

Comment:
This kind of non-blocking write is suspicious to me (at least not idiomatic). Couldn't you simply:
```go
if err != nil && w.errCh != nil {
  select {
    case w.errCh <- err:
        // err consumed by err channel
    default:
        log.Warn("Cannot write error to error channel, it is full", err)
  }
}
```